runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/state/FlinkStateInternals.java

```diff
@@ -1233,4 +1234,121 @@ private void restoreWatermarkHoldsView() throws Exception {
       }
     }
   }
+
+  /** Eagerly create user state to work around https://jira.apache.org/jira/browse/FLINK-12653. */
+  public static class EarlyBinder implements StateBinder {
```

Comment:
Because of my weak understanding of internals, I have to ask. :) 
What is happening here? I see that standard `FlinkStateBinder` use at the end `flinkKeyedStateBackend.getPartitionedState()` but here, it is calling `flinkKeyedStateBackend.getOrCreateKeyedState()`. Why is this OK? Flink API doc did not help me much

Reply:
Feel free to ask anything :) `getPartitionedState` is a wrapper around `getOrCreateKeyedState`. The latter creates a keyed state (as the name suggests). To create a keyed state, you pass in a namespace serializer and a state descriptor. The state descriptor contains a state serializer and a state name. So the state is ready to be used now, but without an active namespace, it cannot be accessed. You can now cast the `State` you received to `InternalKvState` which allows you to set the namespace. The alternative to use `getPartitionedState`, which allows to specify an actual namespace.

Perhaps this is even clearer when looking at the source code:

```java
	public <N, S extends State> S getPartitionedState(
			final N namespace,
			final TypeSerializer<N> namespaceSerializer,
			final StateDescriptor<S, ?> stateDescriptor) throws Exception {
		checkNotNull(namespace, "Namespace");
                /* ... */
		final S state = getOrCreateKeyedState(namespaceSerializer, stateDescriptor);
		final InternalKvState<K, N, ?> kvState = (InternalKvState<K, N, ?>) state;
                /* ... */
		kvState.setCurrentNamespace(namespace);

		return state;
	}
```  

I've omitted some caching code for simplicity.

Reply:
Just to add, calling `getOrCreateKeyedState` is the only way to eagerly create the state because we do not have a namespace available during operator startup.

Reply:
Thanks! 👍